packages/grid/_modules_/grid/hooks/features/useColumnReorder.tsx

```diff
@@ -0,0 +1,150 @@
+import * as React from 'react';
+import { ColDef } from '../../models/colDef';
+import { useLogger } from '../utils';
+import { ApiRef } from '../../models';
+import {
+  DRAGEND,
+  COL_REORDER_START,
+  COL_REORDER_DRAG_OVER,
+  COL_REORDER_DRAG_OVER_HEADER,
+  COL_REORDER_DRAG_ENTER,
+  COL_REORDER_STOP,
+} from '../../constants/eventsConstants';
+import {
+  HEADER_CELL_DROP_ZONE_CSS_CLASS,
+  HEADER_CELL_DRAGGING_CSS_CLASS,
+} from '../../constants/cssClassesConstants';
+
+export interface CursorCoordinates {
+  x: number;
+  y: number;
+}
+
+const CURSOR_MOVE_DIRECTION_LEFT = 'left';
+const CURSOR_MOVE_DIRECTION_RIGHT = 'right';
+
+const reorderColDefArray = (
+  columns: ColDef[],
+  newColIndex: number,
+  oldColIndex: number,
+): ColDef[] => {
+  const columnsClone = [...columns];
```

Comment:
If you want to improve perf:

```suggestion
  const columnsClone = columns.slice();
```

https://www.typescriptlang.org/play?target=1&ts=4.0.2#code/MYewdgzgLgBKA2BXAtpAwvcBTGBeGA2gHQkIqQC6A3EA

Reply:
Done